orchagent/fgnhgorch.cpp

```diff
@@ -965,10 +944,100 @@ bool FgNhgOrch::setNewNhgMembers(FGNextHopGroupEntry &syncd_fg_route_entry, FgNh
+bool FgNhgOrch::isRouteFineGrainedECMP(sai_object_id_t vrf_id, const IpPrefix &ipPrefix, const NextHopGroupKey &nextHops)
 {
     SWSS_LOG_ENTER();
+ 
+    if (!isFineGrainedConfigured || (vrf_id != gVirtualRouterId))
+    {
+        return false;
+    }
 
+    FgNhgEntry *fgNhgEntry = 0;
+    set<NextHopKey> next_hop_set = nextHops.getNextHops();
+    auto prefix_entry = m_fgNhgPrefixes.find(ipPrefix);
+    if (prefix_entry == m_fgNhgPrefixes.end())
+    {
+        for (NextHopKey nhk : next_hop_set)
+        {
+            auto member_entry = m_fgNhgNexthops.find(nhk.ip_address);
+            if (member_entry == m_fgNhgNexthops.end())
+            {
+                return false;
+            }
+
+            if (!fgNhgEntry)
+            {
+                fgNhgEntry = member_entry->second;
+            }
+            else
+            {
+                /* Case where fgNhgEntry is alredy found via previous nexthop
+                 * We validate the it belongs to the same next-hop group set
+                 */
+                if (fgNhgEntry != member_entry->second)
+                {
+                    SWSS_LOG_INFO("FG nh found across different FG_NH groups: %s expected %s, actual %s", 
+                        nhk.to_string().c_str(), fgNhgEntry->fg_nhg_name.c_str(), member_entry->second->fg_nhg_name.c_str());
+                    return false;
+                }
+            }
+        }
+    }
+    return true;
+}
+
+
+bool FgNhgOrch::containsRoute(sai_object_id_t vrf_id, const IpPrefix &ipPrefix)
+{
+    if (!isFineGrainedConfigured || (vrf_id != gVirtualRouterId))
+    {
+        return false;
+    }
+
+    auto it_route_table = m_syncdFGRouteTables.find(vrf_id);
+    if (it_route_table == m_syncdFGRouteTables.end())
+    {
+        return false;
+    }
+
+    auto it_route = it_route_table->second.find(ipPrefix);
+    if (it_route == it_route_table->second.end())
+    {
+        return false;
+    }
+    return true;
+}
+
+
+bool FgNhgOrch::addModifyFgNhg(sai_object_id_t vrf_id, const IpPrefix &ipPrefix, const NextHopGroupKey &nextHops, 
+                                    sai_object_id_t &next_hop_id, bool &prevNhgWasFineGrained)
+{
+    SWSS_LOG_ENTER();
+
+    FgNhgEntry *fgNhgEntry = 0;
+    set<NextHopKey> next_hop_set = nextHops.getNextHops();
+    auto prefix_entry = m_fgNhgPrefixes.find(ipPrefix);
+    if (prefix_entry != m_fgNhgPrefixes.end())
+    {
+        fgNhgEntry = prefix_entry->second;
+    }
+    else
+    {
+        for (NextHopKey nhk : next_hop_set)
+        {
+            auto member_entry = m_fgNhgNexthops.find(nhk.ip_address);
+            if (member_entry == m_fgNhgNexthops.end())
+            {
+                SWSS_LOG_ERROR("fgNhgOrch got a route addition %s:%s for non-configured FG ECMP entry",
+                                    ipPrefix.to_string().c_str(), nextHops.to_string().c_str());
+                return true;
```

Comment:
>return [](start = 16, length = 6)

For early return cases (there are more below), set prevNhgWasFineGrained to false?

Reply:
Made prevNhgWasFineGrained default to true, because for early return cases where addModifyFgNhg returns success, prevNhgWasFineGrained should be true. For early return cases where addModifyFgNhg returns failure, routeorch addRoute will return false directly, so we don't care about prevNhgWasFineGrained on early failure return.

Reply:
I know your code is working. I am arguing about the individual function design. Here you assume the caller takes the responsibility to initialize the output parameter `prevNhgWasFineGrained`. This is not necessary. You can make the decision on its boolean value internally, and output in any cases.

Then outside this function, you don't need to initialize the variable used as an output argument.

---
In reply to: [549525398](https://github.com/Azure/sonic-swss/pull/1565#discussion_r549525398) [](ancestors = 549525398)

Reply:
Yes, it makes sense to have the fn logic take care of the value setting internally. I have set the value in the fn: https://github.com/Azure/sonic-swss/pull/1565/files#diff-076ab0a620bf8e45930d13ac4b8a494eca6d6349b7238bcd52bfeafc1acb7116R1025

The caller is still setting the value to false to prevent an uninitalized var: the value of ```prevNhgWasFineGrained``` is only set when ```isRouteFineGrained``` is true and ```setFgNhg``` is called:
https://github.com/Azure/sonic-swss/pull/1565/files#diff-dd5418e58fe44ea78956f7404758ba1a9eb88d9f1900e29a60a75c79af387480R1256

This uninit var will then be checked in https://github.com/Azure/sonic-swss/pull/1565/files#diff-dd5418e58fe44ea78956f7404758ba1a9eb88d9f1900e29a60a75c79af387480R1445 . So it may be better for readability and to prevent an uninit var issue to have to explicitly set to false.